android/libraries/rib-router-navigator/src/test/kotlin/com/uber/rib/core/RouterAndStateTest.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2023. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.rib.core
+
+import com.google.common.truth.Truth
+import org.junit.Test
+import org.mockito.kotlin.doAnswer
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.whenever
+
+class RouterAndStateTest {
+
+  private val state = mock<RouterNavigatorState>()
+  private val attachTransition = mock<RouterNavigator.AttachTransition<Router<*>, RouterNavigatorState>> {
+    on { buildRouter() } doAnswer { mock() } // Creates new instance per each invocation
+  }
+  private val detachTransition = mock<RouterNavigator.DetachTransition<Router<*>, RouterNavigatorState>>()
+
+  @Test
+  fun router___StateIsNotCacheable_NotForcedToCache___destroyRouterAfterDetach() {
+    whenever(state.isCacheable()).doReturn(false)
+
+    val routerAndState = RouterAndState(
+      state,
+      attachTransition,
+      detachTransition,
+      forceRouterCaching = false
+    )
+
+    val router1 = routerAndState.router
+    routerAndState.detachCallback.onPostDetachFromHost(router1, mock(), false)
+    val router2 = routerAndState.router
+
+    Truth.assertThat(router1).isNotSameAs(router2)
+  }
+
+  @Test
+  fun router___StateIsCacheable_NotForcedToCache___reuseRouterOnNextAttach() {
+    whenever(state.isCacheable()).doReturn(true)
+
+    val routerAndState = RouterAndState(
+      state,
+      attachTransition,
+      detachTransition,
+      forceRouterCaching = false
+    )
+
+    val router1 = routerAndState.router
+    routerAndState.detachCallback.onPostDetachFromHost(router1, mock(), false)
+    val router2 = routerAndState.router
+
+    Truth.assertThat(router1).isSameAs(router2)
+  }
+
+  @Test
+  fun router___StateIsNotCacheable_ForcedToCache___reuseRouterOnNextAttach() {
+    whenever(state.isCacheable()).doReturn(false)
+
+    val routerAndState = RouterAndState(
+      state,
+      attachTransition,
+      detachTransition,
+      forceRouterCaching = true
+    )
+
+    val router1 = routerAndState.router
+    routerAndState.detachCallback.onPostDetachFromHost(router1, mock(), false)
+    val router2 = routerAndState.router
+
+    Truth.assertThat(router1).isSameAs(router2)
+  }
+
+  @Test
+  fun router___StateIsCacheable_ForcedToCache___reuseRouterOnNextAttach() {
```

Comment:
I don't think we use this `__` convention on this lib

Reply:
Make sense, working on change.